### ISSUE TITLE
Invisible widgets that do not display in JSON Editor

### ISSUE THREAD

**Issue by 96LawDawg** [ArnoldSmith86/virtualtabletop#477]
When I was working on updating the chat system, I was doing some editing of the .vtt file offline.  I discovered several dozen widgets that should not be there.  First of all, they should have been deleted by the delete clones routine in the room.  But worse than that, they do not show up in the JSON Editor or the regular editor.  You can only see them in Ghetto or by editing the file offline in a text editor program.  Demo room at https://virtualtabletop.io/wprc.

**Comment by rogerl50** [comment]
It's interesting that all the invisible widgets are cloned from widgets that no longer exist.

**Comment by ArnoldSmith86** [comment]
Yeah, there are some ways to reproduce those kinds of widgets. As an example I looked at `id: t2vs` which has `parent: iwkb`.

When the widget gets added to the room, it doesn't find that parent yet. Instead of adding it immediately, it gets added to a `deferredChildren` array:

https://github.com/ArnoldSmith86/virtualtabletop/blob/1d4fb4c5b4632e59a6a675224e9575bb3b6a9024/client/js/serverstate.js#L18

When another widget is added to the room, this array is checked for any orphaned children and adds the ones that have the new widget as their parent.

There's no check at the end if there are any widgets left though. So any widget without a parent just gets silently ignored.

Other problems (like widgets without `id` or card without `deck` or `cardType`) produce an error message in the browser console when trying to add them.

**Comment by bjalder26** [comment]
I went into the ghetto editor and looked at the ID of the parent of one of them.  Then I created a widget by the same ID and now it shows up in the JSON editor.  I then changed its owner property so it is now visible.  To recreate this issue I added two checker pieces, made one of them the parent of the other, gave that parent the `clonedFrom` property, then used the button to delete the parent.  The child disappears, and is not visible in the JSON editor, but is visible in the Ghetto editor.  In the Ghetto editor, I also gave the child the `clonedFrom` property, then I reuploaded the room and tried to use the button to delete the checker piece.  It did not delete it. 

I could also "bring back" the "partially deleted" checker by renaming another widget with the same ID as the parent that was deleted.

Long story short, it appears to be that children don't get "completely" deleted when their parent is deleted, and then the "partially" deleted widget becomes hard to delete.

**Comment by bjalder26** [comment]
Today I helped someone who had a room that kept crashing and resetting over and over.  This was after they tried to adapt the copy machine from Robart to copy a single tile, and copied a ton (I'm guessing they created a loop). After that, it kept reloading over and over.  Then they tried to wipe the room using the Ghetto Editor.  

I went into the Ghetto editor, found, and deleted these two cards.  Then that behavior stopped. deckA had previously been deleted, and copyMachine had also previously been deleted.  Unfortunately, I didn't think to try and download the file before 'fixing' it, nor did I think to try deleting one at a time.  However, I am hoping this will be a useful clue, and I am guessing this is either a deleted parent issue, or a clonedFrom issue considering one of these cards was cloned from the other. 

{
  "type": "card",
  "cardType": "V2",
  "deck": "deckA",
  "z": 21311,
  "count": false,
  "scrap": false,
  "scale": 1,
  "x": 112,
  "y": 315,
  "clonedFrom": "jmea",
  "id": "e0ls"
}

{
  "type": "card",
  "cardType": "V2",
  "deck": "deckA",
  "z": 21311,
  "count": false,
  "scrap": false,
  "scale": 1,
  "x": 430,
  "y": 131,
  "clonedFrom": "m0al",
  "id": "jmea",
  "parent": "copyMachine"
}

**Comment by ArnoldSmith86** [comment]
Did you try to download thghe rest of the room and putting those two back into it?

**Comment by bjalder26** [comment]
> Did you try to download the rest of the room and putting those two back into it?
No, sorry.  I was in 'fix' mode mentally, and he already overwrote the room before I started thinking about this.  I am trying to recreate the issue though.

**Comment by bjalder26** [comment]
This is the closest I can get tonight.
Made a clone loop. Cloned cards. Deleted the deck. Undeleted cards remain. If you click on the pile, then the room freezes. 
https://virtualtabletop.io/rsuk

**Comment by bjalder26** [comment]
I think this is the best I can do to recreate the issue.  I wanted to create a loop that keeps making cards with a deleted deck to see if it would reach a point where it flickers, but the timer freezes.

https://virtualtabletop.io/xtyr

Move original card to 2nd holder.
cloned twice.
One clone moved to 3rd holder.
The other clone moved back to 1st holder.
Deck cloned, and 1st holder card's deck set to new cloned deck.
Old deck deleted (which should delete the cards in the 2nd and 3rd holders).

Instead, it APPEARS that they all remain.

at 10 seconds the timer moves another card over (you can pause to look at things, but touching the card in holder 1 or 2 sometimes freezes the room)

The timer moving another card into the holder that clones them causes the routine to trigger again, which then freezes the room.

So it appears that interacting with a clone that has a deleted deck freezes the room.
It also appears that triggering the routine again freezes the room (maybe because now a real cloned card is supposed to stack with a card with a deleted deck).

I tried cloning the 2nd holder and making another holder at the bottom that also clones, and things got even weirder when I then placed a card in that holder. 

I'm intentionally trying to break something (which is not a real issue).  However, I'm hoping this helps find an actual issue.
